Update Signal concept (#1311)
temporalio/documentation: docs/concepts/what-is-a-signal.md

id: what-is-a-signal
title: What is a Signal?
sidebar_label: Signal
description: A Signal is an asynchronous request to a Workflow Execution.
tags:
  - signals
  - explanation
---

A Signal is an asynchronous request to a [Workflow Execution](/concepts/what-is-a-workflow-execution).

A Signal delivers data to a running Workflow Execution.
It cannot return data to the caller (use [Queries](/concepts/what-is-a-query) for that).

> The Workflow code that handles a Signal may mutate Workflow state.

Author: @djmagee icymi, waiting on ^

> The Workflow code that handles a Signal may mutate Workflow state.

Reviewer: According to "may" in https://developers.google.com/style/word-list#letter-m, "To convey _possibility_, use _can_ or _might_ instead. To convey _permission_, use _can_ instead."

> The Workflow code that handles a Signal may mutate Workflow state.

Author: I don't like that it doesn't differentiate between the two cans, but okay 😄

> The Workflow code that handles a Signal may mutate Workflow state.

Reviewer: Yeah, Microsoft comes to the same conclusion, albeit in different words: https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/can-may.

A Signal can be sent from a Temporal Client or a Workflow.
When a Signal is sent, it is received by the Cluster and recorded as an Event to the Workflow Execution [Event History](/concepts/what-is-an-event-history).
A successful response from the Cluster means that the Signal has been persisted and will be delivered at least once to the Workflow Execution.[^1]
The next scheduled Workflow Task will contain the Signal Event.

A Signal must include a destination (Namespace & Workflow Id) and name.
It may include a list of arguments.

Signal handlers are Workflow functions that listen for Signals by the Signal name.
Signals are delivered in the order they are received by the Cluster.
If multiple deliveries of a Signal would be a problem for your Workflow, add idempotency logic to your Signal handler that checks for duplicates.

- [How to use Signals](/application-development-guide/#signals)

> [^1]: The Cluster usually deduplicates Signals, but does not guarantee deduplication: there can be two Signal Events (and therefore two deliveries to the Workflow Execution) for a single Signal during shard migration, since the deduping info is only stored in memory.

Reviewer: ```suggestion
[^1]: The Cluster usually deduplicates Signals, but does not guarantee deduplication: During shard migration, two Signal Events (and therefore two deliveries to the Workflow Execution) can occur for a single Signal because the deduping info is stored only in memory.
```
